dotnet/src/SemanticKernel/CoreSkills/ConditionalSkill.cs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        return llmResponse?["variables"]?.Deserialize<string[]>()
+            .Where(v => !string.IsNullOrWhiteSpace(v)).Select(v => v.TrimStart('$')) ?? Enumerable.Empty<string>();
+    }
+
+    private async Task<bool> EvaluateConditionAsync(string ifContent, IEnumerable<string> usedVariables, SKContext context)
+    {
+        var conditionContent = this.ExtractConditionalContent(ifContent);
+
+        context.Variables.Set("IfCondition", conditionContent);
+        context.Variables.Set("ConditionalVariables", this.GetConditionalVariablesFromContext(usedVariables, context.Variables));
+
+        var llmConditionResponse = (await this._evaluateConditionFunction.InvokeAsync(conditionContent, context).ConfigureAwait(false)).ToString();
+
+        var reason = this.GetReason(llmConditionResponse);
+        var error = !Regex.Match(llmConditionResponse.Trim(), @"^(true|false)", RegexOptions.IgnoreCase).Success;
+        if (error)
+        {
+            throw new ConditionException(ConditionException.ErrorCodes.InvalidConditionFormat, reason);
+        }
+
+        return llmConditionResponse.StartsWith("TRUE", StringComparison.OrdinalIgnoreCase);
+    }
+
+    private string ExtractConditionalContent(string ifContent)
+    {
+        XmlDocument xmlDoc = new();
+        xmlDoc.LoadXml("<xml>" + ifContent + "</xml>");
+        XmlNode parentConditionGroupNode = xmlDoc.SelectSingleNode("//if/conditiongroup");
+
+        return parentConditionGroupNode.OuterXml;
+    }
+
+    private string GetConditionalVariablesFromContext(IEnumerable<string> usedVariables, ContextVariables variables)
+    {
+        var conditionalVariables = new StringBuilder();
+
+        var checkNotFoundVariables = usedVariables.Where(u => !variables.ContainsKey(u));
+        if (checkNotFoundVariables.Any())
+        {
+            throw new ConditionException(ConditionException.ErrorCodes.ContextVariablesNotFound, string.Join(", ", checkNotFoundVariables));
+        }
+
+        var existingVariables = variables.Where(v => usedVariables.Contains(v.Key));
+
+        foreach (var v in existingVariables)
+        {
+            // Numeric don't add quotes
+            if (Regex.Match(v.Value, "^[0-9.,]+$").Success)
+            {
+                conditionalVariables.AppendLine($"{v.Key} = {v.Value}");
+            }
+            else
+            {
+                conditionalVariables.AppendLine($"{v.Key} = {JsonSerializer.Serialize(v.Value)}");
+            }
+        }
+
+        return conditionalVariables.ToString();
+    }
+
+    private string? GetReason(string llmResponse)
+    {
+        var hasReasonIndex = llmResponse.IndexOf(ReasonIdentifier, StringComparison.OrdinalIgnoreCase);
+        if (hasReasonIndex > -1)
+        {
+            return llmResponse[(hasReasonIndex + ReasonIdentifier.Length)..].Trim();
+        }
+        return NoReasonMessage;
+    }
+
+    private JsonNode GetValuesFromResponse(string llmResponse)
```

Comment:
Feels like these private methods are good candidates for a separate class, then you can make them `internal` and write unit-tests for them separately.

Reply:
Those are specific to the ConditionalSkill responsability if that is needed in another skill I can move it to a separate class.

I do test this method indirectly mocking a LLMResponse in the unit tests.